File: linux/core/src/NetworkSwitcher.cpp
/* Copyright 2017-2018 All Rights Reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * [Contact]
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0(the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "../inc/NetworkSwitcher.h"

#include "../inc/Core.h"
#include "../inc/SegmentManager.h"
#include "../inc/ServerAdapter.h"

#include "../../common/inc/DebugLog.h"

#include "../../configs/ExpConfig.h"
#include "../../configs/NetworkSwitcherConfig.h"

#include <string.h>
#include <thread>
#include <unistd.h>

using namespace sc;

NetworkSwitcher *NetworkSwitcher::sSingleton = NULL;

SwitchAdapterTransaction *SwitchAdapterTransaction::sOngoing = NULL;
ConnectRequestTransaction *ConnectRequestTransaction::sOngoing = NULL;
ReconnectAdapterTransaction *ReconnectAdapterTransaction::sOngoing = NULL;

void NetworkSwitcher::connect_adapter_by_peer(int adapter_id) {
  NSState state = this->get_state();
  if (state == NSState::kNSStateSwitching) {
    LOG_WARN("connect(adapter %d): FAILED - already switching", adapter_id);
    return;
  }

  this->set_state(NSState::kNSStateSwitching);
  ConnectRequestTransaction::run(adapter_id);
}

void NetworkSwitcher::disconnect_adapter_by_peer(int adapter_id,
                                                 uint32_t last_seq_no_control,
                                                 uint32_t last_seq_no_data) {
  NSState state = this->get_state();
  if (state == NSState::kNSStateSwitching) {
    LOG_WARN("disconnect(adapter %d): FAILED - already switching", adapter_id);
    return;
  }

  Core *core = Core::singleton();
  ServerAdapter *adapter = core->find_adapter_by_id(adapter_id);
  if (adapter == NULL) {
    LOG_WARN("Cannot find adapter %d", (int)adapter_id);
    return;
  }

  this->set_state(NSState::kNSStateSwitching);

  adapter->disconnect_on_peer_command(NULL, last_seq_no_control,
                                      last_seq_no_data);
}

void NetworkSwitcher::sleep_adapter_by_peer(int adapter_id) {
  NSState state = this->get_state();
  if (state == NSState::kNSStateSwitching) {
    LOG_VERB("It's now switching. Cannot connect to adapter %d.", adapter_id);
    return;
  }

  ServerAdapter *adapter = Core::singleton()->get_adapter(adapter_id);
  if (adapter != NULL) {
    this->set_state(NSState::kNSStateSwitching);
    adapter->sleep(NULL, false);
    this->set_state(NSState::kNSStateReady);
  }
}

void NetworkSwitcher::wake_up_adapter_by_peer(int adapter_id) {
  NSState state = this->get_state();
  if (state == NSState::kNSStateSwitching) {
    LOG_VERB("It's now switching. Cannot connect to adapter %d.", adapter_id);
    return;
  }

  ServerAdapter *adapter = Core::singleton()->get_adapter(adapter_id);
  if (adapter != NULL) {
    this->set_state(NSState::kNSStateSwitching);
    adapter->wake_up(NULL, false);
    this->set_state(NSState::kNSStateReady);
  }
}

void NetworkSwitcher::reconnect_adapter(ServerAdapter *adapter,
                                        bool retry_if_already_switching) {
  /* Wait until other adapter is disconnected */
  NSState state = this->get_state();
  if (retry_if_already_switching && state == NSState::kNSStateSwitching) {
    LOG_VERB("It's now switching. Cannot reconnect adapter %s",
             adapter->get_name());
    sleep(1);
    this->reconnect_adapter(adapter, retry_if_already_switching);
    return;
  }

  /* Start to reconnect the adapter */
  this->set_state(NSState::kNSStateSwitching);
  ReconnectAdapterTransaction::run(adapter);
}

bool NetworkSwitcher::switch_adapters(int prev_index, int next_index) {
  // Switch Step 0
  this->set_state(NSState::kNSStateSwitching);
  return SwitchAdapterTransaction::run(prev_index, next_index);
}

void NetworkSwitcher::done_switch() {
  LOG_IMP("Switch adapter end!");
  NSState state = this->get_state();
  switch (state) {
  case NSState::kNSStateSwitching:
    this->set_state(NSState::kNSStateReady);
    break;
  case NSState::kNSStateReady:
    break;
  }
}

bool SwitchAdapterTransaction::run(int prev_index, int next_index) {
  if (sOngoing == NULL) {
    sOngoing = new SwitchAdapterTransaction(prev_index, next_index);
    LOG_VERB("Switch (%d->%d): START", prev_index, next_index);
    sOngoing->start();
    return true;
  } else {
    LOG_WARN("Already starting core");
    NetworkSwitcher *switcher = NetworkSwitcher::singleton();
    switcher->done_switch();
    return false;
  }
}

void SwitchAdapterTransaction::done(bool is_success) {
  // SWITCH TIME: disconnect end
  this->setDisconnectEndTS();
  uint64_t totalLatency = this->getTotalSwitchLatency();
  uint64_t connectLatency = this->getConnectLatency();
  uint64_t sleepLatency = this->getSleepLatency();
  uint64_t disconnectLatency = this->getDisconnectLatency();

  if (is_success) {
    LOG_IMP("Switch (%d->%d): SUCCESS (%lluus = %lluus + %lluus + %lluus)",
            sOngoing->mPrevIndex, sOngoing->mNextIndex, totalLatency,
            connectLatency, sleepLatency, disconnectLatency);
    Core::singleton()->set_active_adapter_index(sOngoing->mNextIndex);
  } else {
    LOG_ERR("Switch (%d->%d): FAILED (%lluus = %lluus + %lluus + %lluus)",
            sOngoing->mPrevIndex, sOngoing->mNextIndex, totalLatency,
            connectLatency, sleepLatency, disconnectLatency);
  }
  NetworkSwitcher::singleton()->done_switch();
  sOngoing = NULL;
}

void SwitchAdapterTransaction::start(void) {
  // Switch Step 0, 1-a
  Core *core = Core::singleton();

  // SWITCH TIME: connect start
  this->setConnectStartTS();

  ServerAdapter *next_adapter = core->get_adapter(this->mNextIndex);
  if (next_adapter == NULL) {
    LOG_ERR("Connecting next adapter is failed");
    this->done(false);
    return;
  }

  ServerAdapterState adapter_state = next_adapter->get_state();
  if (adapter_state == ServerAdapterState::kActive) {
    LOG_WARN("Next adapter is already active");
    SwitchAdapterTransaction::connect_next_adapter_callback(NULL, true);
    return;
  } else if (adapter_state == ServerAdapterState::kConnecting ||
             adapter_state == ServerAdapterState::kDisconnecting ||
             adapter_state == ServerAdapterState::kGoingSleeping ||
             adapter_state == ServerAdapterState::kWakingUp) {
    LOG_ERR("Switching is already in progress");
    this->done(false);
    return;
  }

  LOG_VERB("Switch (%d->%d): STEP 1. %s Adapter ON", sOngoing->mPrevIndex,
           sOngoing->mNextIndex, next_adapter->get_name());

// Connect or wake up the next adapter
#ifdef EXP_DONT_SEND_CONNECT_CONTROL_MESSAGE
  // For debugging
  next_adapter->connect_or_wake_up(
      SwitchAdapterTransaction::connect_next_adapter_callback, false);
#else
  next_adapter->connect_or_wake_up(
      SwitchAdapterTransaction::connect_next_adapter_callback, true);
#endif
}

void SwitchAdapterTransaction::connect_next_adapter_callback(
    ServerAdapter *adapter, bool is_success) {
  // Switch Step 1-b, 2-a
  Core *core = Core::singleton();
  NetworkSwitcher *switcher = NetworkSwitcher::singleton();

  // SWITCH TIME: sleep start
  sOngoing->setSleepStartTS();

  // Check if the connection result is successful
  if (!is_success) {
    LOG_ERR("Connecting next adapter is failed");
    sOngoing->done(false);
    return;
  }

  // Get previous adapter
  ServerAdapter *prev_adapter = core->get_adapter(sOngoing->mPrevIndex);
  if (prev_adapter == NULL) {
    LOG_ERR("Disconnecting previous adapter is failed");
    sOngoing->done(false);
    return;
  }

  // Check previous adapter's state
  ServerAdapterState adapter_state = prev_adapter->get_state();
  if (adapter_state == ServerAdapterState::kDisconnected) {
    LOG_WARN("Prev adapter is already disconnected");
    // Proceed to Step 4
    sOngoing->done(true);
    return;
  } else if (adapter_state == ServerAdapterState::kSleeping ||
             adapter_state == ServerAdapterState::kGoingSleeping) {
    LOG_WARN("Prev adapter is already sleeping or going to sleeping");
    // Proceed to Step 3-a
    sOngoing->sleep_prev_adapter_callback(NULL, true);
  } else if (adapter_state == ServerAdapterState::kConnecting ||
             adapter_state == ServerAdapterState::kDisconnecting ||
             adapter_state == ServerAdapterState::kWakingUp) {
    // Failure
    LOG_ERR("Switching is already in progress");
    sOngoing->done(false);
    return;
  }

  LOG_VERB("Switch (%d->%d): STEP 2. %s Adapter SLEEP", sOngoing->mPrevIndex,
           sOngoing->mNextIndex, prev_adapter->get_name());

  // Disconnect or sleep previous adapter
  // TODO: Actually, it needs sleep request. For now, since sleep does not
  // support ACK message, sleep request message makes out-of-order of sleep()
  // and disconnect().
  prev_adapter->sleep(SwitchAdapterTransaction::sleep_prev_adapter_callback,
                      false);
}

void SwitchAdapterTransaction::sleep_prev_adapter_callback(
    ServerAdapter *adapter, bool is_success) {
  // Switch Step 2-b, 3-a
  Core *core = Core::singleton();
  NetworkSwitcher *switcher = NetworkSwitcher::singleton();

  // SWITCH TIME: disconnect start
  sOngoing->setDisconnectStartTS();

  if (!is_success) {
    LOG_ERR("Sleeping next adapter is failed");
    sOngoing->done(false);
    return;
  }

  ServerAdapter *prev_adapter = core->get_adapter(sOngoing->mPrevIndex);
  if (prev_adapter == NULL) {
    LOG_ERR("Sleeping previous adapter is failed");
    sOngoing->done(false);
    return;
  }

  ServerAdapterState adapter_state = prev_adapter->get_state();
  if (adapter_state == ServerAdapterState::kDisconnected) {
    // Proceed to Step 4
    LOG_WARN("Prev adapter is already disconnected");
    sOngoing->done(true);
    return;
  } else if (adapter_state == ServerAdapterState::kGoingSleeping) {
    // Wait until the adapter become sleeping state
    LOG_VERB("Wait for going to sleeping");
    do {
      adapter_state = prev_adapter->get_state();
      ::sleep(1);
    } while (adapter_state != ServerAdapterState::kSleeping);
  } else if (adapter_state == ServerAdapterState::kConnecting ||
             adapter_state == ServerAdapterState::kDisconnecting ||
             adapter_state == ServerAdapterState::kWakingUp) {
    // Failure
    LOG_ERR("Switching is already in progress");
    sOngoing->done(false);
    return;
  }

  LOG_VERB("Switch (%d->%d): STEP 3. %s Adapter OFF", sOngoing->mPrevIndex,
           sOngoing->mNextIndex, prev_adapter->get_name());

  prev_adapter->disconnect_on_command(
      SwitchAdapterTransaction::disconnect_prev_adapter_callback);
}

void SwitchAdapterTransaction::disconnect_prev_adapter_callback(
    ServerAdapter *adapter, bool is_success) {
  // Switch Step 3-b, 4
  if (!is_success) {
    LOG_ERR("Disconnecting prev adapter is failed");
    sOngoing->done(false);
    return;
  }

  // Switching adapter is successfully done!
  sOngoing->done(true);
}

bool ConnectRequestTransaction::run(int adapter_id) {
  if (sOngoing == NULL) {
    sOngoing = new ConnectRequestTransaction(adapter_id);
    sOngoing->start();
    return true;
  } else {
    LOG_WARN("Already starting core");
    NetworkSwitcher *switcher = NetworkSwitcher::singleton();
    switcher->done_switch();
    return false;
  }
}

void ConnectRequestTransaction::done() {
  NetworkSwitcher *switcher = NetworkSwitcher::singleton();
  switcher->done_switch();
  sOngoing = NULL;
}

// Connect Request
bool ConnectRequestTransaction::start() {
  // Connect requested adapter
  ServerAdapter *adapter =
      Core::singleton()->find_adapter_by_id(this->mAdapterId);
  if (adapter == NULL) {
    LOG_ERR("Connecting requested adapter is failed");
    this->done();
    return false;
  }
  adapter->connect(ConnectRequestTransaction::connect_callback, false);
  return true;
}

void ConnectRequestTransaction::connect_callback(ServerAdapter *adapter,
                                                 bool is_success) {
  if (!is_success) {
    LOG_ERR("Connecting requested adapter is failed");
    sOngoing->done();
    return;
  }
  LOG_VERB("Connecting requested adapter is done");
  sOngoing->done();
}

bool ReconnectAdapterTransaction::run(ServerAdapter *targetAdapter) {
  if (sOngoing == NULL) {
    sOngoing = new ReconnectAdapterTransaction(targetAdapter);
    sOngoing->start();
    return true;
  } else {
    LOG_WARN("Already starting core");
    NetworkSwitcher *switcher = NetworkSwitcher::singleton();
    switcher->done_switch();
    return false;
  }
}

void ReconnectAdapterTransaction::done(bool require_restart) {
  sOngoing = NULL;
  if (require_restart) {
    ReconnectAdapterTransaction::start();
  } else {
    sleep(1);

    NetworkSwitcher *switcher = NetworkSwitcher::singleton();
    switcher->done_switch();
  }
}

// Reconnect Adapter
bool ReconnectAdapterTransaction::start() {
  // disconnect adapter
  if (this->mTargetAdapter == NULL) {
    LOG_ERR("Reconnecting adapter is failed: retry");
    this->done(true);
    return false;
  }
  this->mTargetAdapter->disconnect_on_failure(
      ReconnectAdapterTransaction::disconnect_callback);
  return true;
}

void ReconnectAdapterTransaction::disconnect_callback(ServerAdapter *adapter,
                                                      bool is_success) {
  if (!is_success) {
    LOG_ERR("Reconnecting adapter is failed: retry");
    sOngoing->done(true);
    return;
  }
  // connect adapter
  if (sOngoing->mTargetAdapter == NULL) {
    LOG_ERR("Reconnecting adapter is failed: retry");
    sOngoing->done(true);
    return;
  }
  sOngoing->mTargetAdapter->connect(
      ReconnectAdapterTransaction::disconnect_callback, false);
}

void ReconnectAdapterTransaction::connect_callback(ServerAdapter *adapter,
                                                   bool is_success) {
  if (!is_success) {
    LOG_ERR("Reconnecting adapter is failed 6: retry");
    sOngoing->done(true);
    return;
  }
  LOG_VERB("Reconnecting adapter is done");
  NetworkSwitcher::singleton()->done_switch();
}

File: framework/selective-connection/android/app/src/main/java/com/redcarrottt/sc/internal/NetworkSwitcher.java
package com.redcarrottt.sc.internal;

/* Copyright (c) 2017-2018. All rights reserved.
 *  Gyeonghwan Hong (redcarrottt@gmail.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import com.redcarrottt.testapp.Logger;

public class NetworkSwitcher {
    private static final String kTag = "NetworkSwitcher";

    // Connect adapter command.
    // It is called by peer through Core.
    void connectAdapterByPeer(int adapterId) {
        int state = this.getState();
        if (state == State.kSwitching) {
            Logger.VERB(kTag, "It's now switching. Cannot connect to adapter " + adapterId);
            return;
        }
        this.setState(State.kSwitching);
        runConnectRequestTx(adapterId);
    }

    // Disconnect adapter command.
    // It is called by peer through Core.
    void disconnectAdapterByPeer(int adapterId, int finalSeqNoControl, int finalSeqNoData) {
        int state = this.getState();
        if (state == State.kSwitching) {
            Logger.VERB(kTag, "It's now switching. Cannot disconnect to adapter " + adapterId);
            return;
        }

        Core core = Core.singleton();
        ClientAdapter adapter = core.findAdapterById(adapterId);
        if (adapter == null) {
            Logger.WARN(kTag, "Cannot find adapter " + adapterId);
            return;
        }

        this.setState(State.kSwitching);

        adapter.disconnectOnPeerCommand(onResultDisconnectAdapterByPeer, finalSeqNoControl, finalSeqNoData);
    }

    ClientAdapter.DisconnectResultListener onResultDisconnectAdapterByPeer = new ClientAdapter
            .DisconnectResultListener() {
        @Override
        public void onDisconnectResult(boolean isSuccess) {
            NetworkSwitcher switcher = NetworkSwitcher.singleton();
            switcher.doneSwitch();
        }
    };

    // Sleep adapter command.
    // It is called by peer through Core.
    void sleepAdapterByPeer(int adapterId) {
        int state = this.getState();
        if (state == State.kSwitching) {
            Logger.VERB(kTag, "It's now switching. Cannot sleep to adapter " + adapterId);
            return;
        }
        this.setState(State.kSwitching);
        ClientAdapter adapter = Core.singleton().findAdapterById(adapterId);
        adapter.sleep(false);
        this.setState(State.kReady);
    }

    // Wake up adapter command.
    // It is called by peer through Core.
    void wakeUpAdapterByPeer(int adapterId) {
        int state = this.getState();
        if (state == State.kSwitching) {
            Logger.VERB(kTag, "It's now switching. Cannot sleep to adapter " + adapterId);
            return;
        }
        this.setState(State.kSwitching);
        ClientAdapter adapter = Core.singleton().findAdapterById(adapterId);
        adapter.wakeUp(false);
        this.setState(State.kReady);
    }

    // Reconnect adapter command.
    // It is called by Core.
    void reconnectAdapter(ClientAdapter adapter) {
        // If it is disconnecting on purpose, do not reconnect it.
        if (adapter.isDisconnectingOnPurpose()) {
            return;
        }

        int state = this.getState();
        if (state == State.kSwitching) {
            Logger.VERB(kTag, adapter.getName() + ": It's now switching. Cannot reconnect " +
                    "adapter" + ".");
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            this.reconnectAdapter(adapter);
            return;
        }
        this.setState(State.kSwitching);
        runReconnectAdapterTx(adapter);
    }

    // Notification of switch done event
    private void doneSwitch() {
        int state = this.getState();
        switch (state) {
            case State.kSwitching:
                this.setState(State.kReady);
                break;
            case State.kReady:
                break;
        }
    }

    // State Getter/Setter
    @SuppressWarnings("SynchronizeOnNonFinalField")
    private int getState() {
        int state;
        synchronized (this.mState) {
            state = this.mState;
        }
        return state;
    }

    @SuppressWarnings("SynchronizeOnNonFinalField")
    private void setState(int newState) {
        synchronized (this.mState) {
            this.mState = newState;
        }
    }

    class State {
        public static final int kReady = 0;
        public static final int kSwitching = 1;
    }

    // State
    private Integer mState;

    // Singleton
    private static NetworkSwitcher sSingleton;

    public static NetworkSwitcher singleton() {
        if (sSingleton == null) {
            sSingleton = new NetworkSwitcher();
        }
        return sSingleton;
    }

    // Constructor
    private NetworkSwitcher() {
        // Since there is no NetworkSwitcher for Android implementation,
        // NetworkSwitcher's initial state goes to Running at beginning.
        this.mState = State.kReady;
    }

    // Transactions
    // ----------------------------------------------------------------
    // Singleton runner
    private static boolean runConnectRequestTx(int adapterId) {
        if (sOngoingConnectRequest == null) {
            sOngoingConnectRequest = new ConnectRequestTransaction(adapterId);
            sOngoingConnectRequest.start();
            return true;
        } else {
            Logger.WARN(kTag, "Already connecting");
            return false;
        }
    }

    private static void doneConnectRequestTx(boolean isSuccess) {
        sOngoingConnectRequest = null;
        if (!isSuccess) {
            Logger.WARN(kTag, "Connection request failed");
        }
        NetworkSwitcher.singleton().doneSwitch();
        sOngoingConnectRequest = null;
    }

    private static void runReconnectAdapterTx(ClientAdapter adapter) {
        if (sOngoingReconnectAdapter == null) {
            sOngoingReconnectAdapter = new ReconnectAdapterTransaction(adapter);
            sOngoingReconnectAdapter.start();
        } else {
            Logger.WARN(kTag, "Already stopping core");
        }
    }

    private static void restartReconnectAdapterTx(ClientAdapter adapter) {
        try {
            Thread.sleep(1000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        sOngoingReconnectAdapter = null;
        runReconnectAdapterTx(adapter);
    }

    private static ConnectRequestTransaction sOngoingConnectRequest = null;
    private static ReconnectAdapterTransaction sOngoingReconnectAdapter = null;

    private static class ConnectRequestTransaction {
        private ConnectRequestTransaction(int adapterId) {
            this.mAdapterId = adapterId;
            this.onConnectAdapter = new OnConnectAdapter();
        }

        void start() {
            ClientAdapter adapter = Core.singleton().findAdapterById(this.mAdapterId);
            if (adapter == null) {
                Logger.ERR(kTag, "Connecting requested adapter is failed");
                doneConnectRequestTx(false);
                return;
            }
            adapter.connect(onConnectAdapter, false);
        }

        private OnConnectAdapter onConnectAdapter;

        class OnConnectAdapter implements ClientAdapter.ConnectResultListener {
            @Override
            public void onConnectResult(boolean isSuccess) {
                if (!isSuccess) {
                    Logger.ERR(kTag, "Connecting requested adapter is failed");
                    doneConnectRequestTx(false);
                    return;
                }
                Logger.VERB(kTag, "Connecting requested adapter is done");
                doneConnectRequestTx(true);
            }
        }

        private int mAdapterId;
    }

    private static class ReconnectAdapterTransaction {
        private ReconnectAdapterTransaction(ClientAdapter targetAdapter) {
            this.onConnectAdapter = new OnConnectAdapter();
            this.onDisconnectAdapter = new OnDisconnectAdapter();
            this.mTargetAdapter = targetAdapter;
        }

        private ClientAdapter mTargetAdapter;

        void start() {
            if (this.mTargetAdapter == null) {
                Logger.ERR(kTag, "Reconnecting adapter is failed: retry");
                restartReconnectAdapterTx(mTargetAdapter);
                return;
            }
            this.mTargetAdapter.disconnectOnFailure(onDisconnectAdapter);
        }

        private OnDisconnectAdapter onDisconnectAdapter;

        class OnDisconnectAdapter implements ClientAdapter.DisconnectResultListener {
            @Override
            public void onDisconnectResult(boolean isSuccess) {
                if (!isSuccess) {
                    Logger.ERR(kTag, "Reconnecting adapter is failed: retry");
                    restartReconnectAdapterTx(mTargetAdapter);
                    return;
                }
                if (mTargetAdapter == null) {
                    Logger.ERR(kTag, "Reconnecting adapter is failed: retry");
                    restartReconnectAdapterTx(mTargetAdapter);
                    return;
                }

                mTargetAdapter.connect(onConnectAdapter, false);
            }
        }

        private OnConnectAdapter onConnectAdapter;

        class OnConnectAdapter implements ClientAdapter.ConnectResultListener {
            @Override
            public void onConnectResult(boolean isSuccess) {
                if (!isSuccess) {
                    Logger.ERR(kTag, "Reconnecting adapter is failed: retry");
                    restartReconnectAdapterTx(mTargetAdapter);
                    return;
                }

                Logger.VERB(kTag, "Reconnecting adapter is done");
                NetworkSwitcher.singleton().doneSwitch();
            }
        }
    }
}